### PR TITLE
refactor: [Noon] adding new field max_amount to mandate request

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Resolves  [#3844](https://github.com/juspay/hyperswitch-cloud/issues/3844)

_Note
1.While connector create ensure `test_mode` is set to true
2.Ensure the currency passed is `AED` and country is `AE`_

## How did you test it?
1.Create a CIT with noon
 with 
```
  "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    }
```
Screenshot 2024-01-12 at 2 16 26 PM

</n>
without order_category it should throw an error

Screenshot 2024-01-12 at 2 16 44 PM

2. Make a MIT

Screenshot 2024-01-12 at 2 02 35 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/noon/transformers.rs [hunk 1]
```diff
@@ -330,17 +336,37 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for NoonPaymentsRequest {
...
+        let (subscription, tokenize_c_c) = match item.request.setup_future_usage.is_some() {
+            true => {
+                let mandate_data = item.request.get_setup_mandate_details().ok_or(
+                    errors::ConnectorError::MissingRequiredField {
+                        field_name: "setup_mandate_details",
+                    },
+                )?;
+                let max_amount = match &mandate_data.mandate_type {
+                    Some(data_models::mandates::MandateDataType::SingleUse(mandate)) => {
+                        mandate.amount.to_string()
```
Comment (SamraatBansal) [inline crates/router/src/connector/noon/transformers.rs:348]: Amount conversion is missing

File: crates/router/src/connector/noon/transformers.rs [hunk 2]
```diff
@@ -330,17 +336,37 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for NoonPaymentsRequest {
...
+                    }
+                    Some(data_models::mandates::MandateDataType::MultiUse(Some(mandate))) => {
+                        mandate.amount.to_string()
```
Comment (SamraatBansal) [inline crates/router/src/connector/noon/transformers.rs:351]: Amount conversion is missing

File: crates/router/src/connector/noon/transformers.rs [hunk 3]
```diff
@@ -330,17 +336,37 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for NoonPaymentsRequest {
...
-                NoonSubscriptionData {
-                    subscription_type: NoonSubscriptionType::Unscheduled,
-                    name: name.clone(),
-                },
-                true,
-            )) {
-                Some((a, b)) => (Some(a), Some(b)),
-                None => (None, None),
-            };
+        let (subscription, tokenize_c_c) = match item.request.setup_future_usage.is_some() {
```
Comment (SamraatBansal) [inline crates/router/src/connector/noon/transformers.rs:339]: `match` statement can be refactored from `setup_future_usage` as we can use `setup_mandate_details` like used in stripe.

**Review by SamraatBansal — CHANGES_REQUESTED** [review]
The testing should be consistent across local DB, the response and the noon dashboard so things like amount validation don't get missed

**Comment by AkshayaFoiger** [comment]
> The testing should be consistent across local DB, the response and the noon dashboard so things like amount validation don't get missed

Tested, mandate amount remains consistent in the local DB and dashboard

**Review by SamraatBansal — APPROVED** [review]

**Review by srujanchikke — DISMISSED** [review]

File: crates/router/src/connector/noon/transformers.rs [hunk 4]
```diff
@@ -296,7 +295,11 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for NoonPaymentsRequest {
                     }
                 }?,
                 Some(item.request.currency),
-                item.request.order_category.clone(),
+                Some(item.request.order_category.clone().ok_or(
+                    errors::ConnectorError::MissingRequiredField {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:299]: If the field is non optional why we have defined category in NoonOrder as Option?

**Review by SamraatBansal — DISMISSED** [review]

File: crates/router/src/connector/noon/transformers.rs [hunk 5]
```diff
@@ -330,17 +333,33 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for NoonPaymentsRequest {
...
+        let subscription = item
+            .request
+            .get_setup_mandate_details()
+            .map(|mandate_data| {
+                let max_amount = match &mandate_data.mandate_type {
+                    Some(data_models::mandates::MandateDataType::SingleUse(mandate))
+                    | Some(data_models::mandates::MandateDataType::MultiUse(Some(mandate))) => {
+                        conn_utils::to_currency_base_unit(mandate.amount, item.request.currency)
+                    }
+                    _ => Err(errors::ConnectorError::MissingRequiredField {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:345]: Why do we need to throw error if mandate_data.mandate_type is None? We can simply return None

Comment (AkshayaFoiger) [reply crates/router/src/connector/noon/transformers.rs:345]: When setup_mandate_details is present, our system will create a mandate id. But to create a mandate payment with noon`max_amount` is a required field, hence we are throwing an error here

**Review by ArjunKarthik — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**